src/order_item.rs

```diff
@@ -0,0 +1,220 @@
+        let definitely_skip = skip.unwrap_or(0);
+        let definitely_first = first.unwrap_or(usize::MAX);
+        let discounts_part: Vec<Discount> = discounts
+            .into_iter()
+            .skip(definitely_skip)
+            .take(definitely_first)
+            .collect();
+        let has_next_page = total_count > discounts_part.len() + definitely_skip;
+        Ok(DiscountConnection {
+            nodes: discounts_part,
+            has_next_page,
+            total_count: total_count as u64,
+        })
+    }
+}
+
+impl PartialOrd for OrderItem {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self._id.partial_cmp(&other._id)
+    }
+}
+
+impl Ord for OrderItem {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self._id.cmp(&other._id)
+    }
+}
+
+/// Sorts vector of discounts according to BaseOrder.
+///
+/// * `discounts` - Vector of discounts to sort.
+/// * `order_by` - Specifies order of sorted result.
+fn sort_discounts(discounts: &mut Vec<Discount>, order_by: Option<CommonOrderInput>) {
+    let comparator: fn(&Discount, &Discount) -> bool =
+        match order_by.unwrap_or_default().direction.unwrap_or_default() {
+            OrderDirection::Asc => |x, y| x < y,
+            OrderDirection::Desc => |x, y| x > y,
+        };
+    discounts.sort_by(|x, y| match comparator(x, y) {
+        true => Ordering::Less,
+        false => Ordering::Greater,
+    });
+}
+
+/// Applies fees and discounts to calculate the compensatable amount of an OrderItem.
+fn calculate_compensatable_amount(
+    product_variant_version: ProductVariantVersion,
+    internal_discounts: &BTreeSet<Discount>,
+    shipment_fee: u64,
+) -> u64 {
+    let undiscounted_price = product_variant_version.price as f64;
+    let maybe_max_discountable_amount = internal_discounts
+        .iter()
+        .map(|d| d.max_discountable_amount)
+        .min()
+        .and_then(|v| Some(v as f64));
+    let mut discounted_price = internal_discounts
+        .iter()
+        .fold(undiscounted_price, |prev_price, discount| {
+            prev_price * discount.discount
+        });
+    if let Some(max_discountable_amount) = maybe_max_discountable_amount {
+        if max_discountable_amount > discounted_price - undiscounted_price {
+            discounted_price = undiscounted_price - max_discountable_amount;
+        }
+    }
+    let total_price = discounted_price as u64 + shipment_fee;
+    total_price
+}
+
+/// Describes DTO of an OrderItem of an Order.
+///
+/// `product_item` is set to None as long as `OrderStatus::Pending`.
+/// Must contain a ProductItem when `OrderStatus::Placed` or `OrderStatus::Rejected`.
+#[derive(Debug, Serialize)]
+pub struct OrderItemDTO {
+    /// OrderItem UUID.
+    pub id: Uuid,
+    /// Timestamp when OrderItem was created.
+    pub created_at: DateTime,
+    /// UUID of product item associated with OrderItem.
+    pub product_item_id: Option<Uuid>,
```

Comment:
shouldn't this be called `order_item_id`?

Reply:
Maybe i can rename it to inventory item to make this clearer, the order item id is set above

Reply:
Oh I'm blind, pls ignore

Reply:
Ok, but then a new issue
Doesn't each product item have its own id? Meaning if there are 20 items in the order item, wouldn't you need 20 product item IDs?